packages/flutter_tools/lib/src/flutter_manifest.dart

```diff
@@ -640,17 +596,48 @@ void _validateDeferredComponents(MapEntry<Object?, Object?> kvp, List<String> er
         }
       }
       if (valueMap.containsKey('assets')) {
-        final Object? assets = valueMap['assets'];
-        if (assets is! YamlList) {
-          errors.add('Expected "assets" to be a list, but got $assets (${assets.runtimeType}).');
-        } else {
-          _validateListType<String>(assets, errors, '"assets" key in the $i element of "${kvp.key}"', 'file paths');
-        }
+        errors.addAll(_validateAssets(valueMap['assets']));
       }
     }
   }
 }
 
+List<String> _validateAssets(Object? yaml) {
+  final (_, List<String> errors) = _computeAssetsSafe(yaml);
+  return errors;
+}
+
+(List<AssetsEntry>, List<String> errors) _computeAssetsSafe(Object? yaml) {
```

Comment:
if I'm reading this code correctly, When we call `FlutterManifest._createFromYaml()`, we will be parsing these assets twice, once from calling `_validate()` -> `_validateFlutter()` -> `_validateAssets()` -> `computeAssetsSafe()` and later, from `FlutterManifest.assets` -> `_computeAssets()` -> `computeAssetsSafe()`.

And if you have deferred components we will parse it many more times. Can we call this function *once* from `FlutterManifest._createFromYaml()` and re-use both the erros and the `List<AssetEntry>`?

Reply:
Yes, this is correct, and this pattern is found throughout the `FlutterManifest` class (see the `fonts`, `shaders`, and `models` fields). I'd be happy to attempt to refactor this class to parse and store everything upfront, but this would be a substantial refactoring. Because of this, I elected not to break this pattern in this PR. Do you believe I should attempt this refactoring now?

Reply:
@christopherfujino friendly ping

Reply:
Done in [add tech debt note about redundant parsing](https://github.com/flutter/flutter/pull/132985/commits/00a5a87bc727274a037bdb4ab10f72a08c04aceb)